### ISSUE TITLE
[LLVM] Possible heap corruption in managed mode while trying to format a string using fmt

### ISSUE THREAD

**Issue by cagriuslu-adyen** [oracle/graal#4137]
**Describe GraalVM and your environment :**
```
GraalVM-21.3.0 and GraalVM-21.2.0
Enterprise Edition
JDK11
macOS Big Sur
amd64
Output of `java -Xinternalversion`: `Java HotSpot(TM) 64-Bit Server VM (11.0.12+8-LTS-jvmci-21.2-b06) for bsd-amd64 JRE (11.0.12+8-LTS-jvmci-21.2-b06), built on Jul  1 2021 04:44:57 by "graal2" with gcc 4.2.1 Compatible Apple LLVM 11.0.0 (clang-1100.0.33.17)`
```

**Have you verified this issue still happens when using the latest snapshot?**
Not verified, but it happens with the latest release (21.3.0) and 21.2.0.

**Describe the issue**
Trying to format a string using fmt library causes a `Allocation of native memory is not possible in managed mode` error while running under GraalVM. Suspecting of heap corruption because adding bunch of debug printf's before the crash hides the issue. Memory analysis tools don't report a problem with the library itself (valgrind, etc.) outside of GraalVM.

Attached a small reproducable code that has build instructions inside. Bundled LLVM in the 21.2.0 is used. We've had some problems with the bundled LLVM in the 21.3.0 version, thus we avoided it. (see https://github.com/oracle/graal/issues/4059)

[graalvm-fmt-minimal-example.zip](https://github.com/oracle/graal/files/7735300/graalvm-fmt-minimal-example.zip)

**Comment by rschatz** [comment]
I'm looking into it.

Note that it's unlikely this is really a heap corruption. The error message indicates that something wants to access the native heap, and our safeguards against this kick in as expected.

Usually this happens when pointer arithmetic goes wrong, for example if the C code is trying to do weird stuff with pointer values. Sometimes this is actually undefined behavior in C that just happens to work randomly in native mode, but with managed mode it breaks.

More likely it's the C compiler that is messing around with the code through optimizations. This would explain your observed behavior: The printf statements are side-effects that might prevent that compiler optimization.

If the theory with the compiler optimization is true, then I'd still consider this a bug on our side. We should support all code patterns that can be produced by LLVM with the default flags.

**Comment by cagriuslu-adyen** [comment]
Thanks for the explanation, I'll try to see what happens when optimizations are off while you look into the exception.

The bundled LLVM, I assumed it would ignore the optimization and debug flags and use the most appropriate flags. Is this a wrong assumption?

**Comment by cagriuslu-adyen** [comment]
When `-O1 -g` is used, the issue disappears, thank you very much. Feel free to keep the issue open and investigate.

**Comment by rschatz** [comment]
> The bundled LLVM, I assumed it would ignore the optimization and debug flags and use the most appropriate flags.

No, the bundled clang wrapper will just add flags, it will never remove flags that you specify manually. If you don't specify anything, it defaults to `-O1 -g`, which is the flag combination we recommend.

In general we try to be compatible with all possible optimization flags, but for the higher levels its tricky, and most of the time it's counterproductive, the two optimizing compilers (LLVM and GraalVM) tend to step on each others toes, making the end result less optimal.

> When `-O1 -g` is used, the issue disappears, thank you very much. Feel free to keep the issue open and investigate.

Awesome. Thanks for checking. I will leave this issue open, I would still like to fix this, but probably not in `22.0`, since we're already past the feature freeze, and the actual fix is a bit more complicated (the problematic bitcode involves two pointer values packed into an `i128` integer ;) ).
